Removed noop with proper implementations

morses-code opened the pull request:
Check code makes sense and that tests are passing.

- Removing noop implementations and using correct ones.

morses-code committed:
Removed noop with proper implementations

nshumoogum commented on api/api.go:
```diff
@@ -22,12 +25,26 @@ func Setup(ctx context.Context, r *mux.Router, permissions AuthHandler) *API {
-	r.HandleFunc("/sessions", permissions.Require(create, CreateSessionHandlerFunc(nopSession, nopCache))).Methods("POST")
-	r.HandleFunc("/sessions/{ID}", GetByIDSessionHandlerFunc(nopCache, mux.Vars)).Methods("GET")
-	r.HandleFunc("/sessions", permissions.Require(delete, DeleteAllSessionsHandlerFunc(nopCache))).Methods("DELETE")
+	cfg, err := config.Get()
```
Setting up session and then cache should be in service.Run() function

Tip: Never call `config.Get()` twice in an application, if someone updates the configuration different parts of the app will retrieve different configuration values, even worse if these are then stored as new parameters. This can lead to querky behaviour at best and data issues or app failure at worst.

nshumoogum replied:
I would move lines 30 to 44 into `service.Run()` func in `service/service.go`, then pass the session and cache to the api.

morses-code committed:
Updated handler and api signatures

- Updated handler signatures to remove session
- Updated tests, no longer using session mock
- Updated api.Setup() to include elasticache client

morses-code committed:
Removing redundant files

nshumoogum commented on api/handlers_test.go:
```diff
@@ -63,27 +53,17 @@ func TestCreateSessionHandlerFunc(t *testing.T) {
-				So(resp.Body.String(), ShouldEqual, string(b))
 				So(resp.Code, ShouldEqual, 201)
-				So(mockSession.UpdateCalls(), ShouldHaveLength, 1)
-				So(mockSession.UpdateCalls()[0].Email, ShouldEqual, "test@test.com")
```
So I would have checked that you still get these values. I've written the assertions locally and found a bug with the session struct, json tag for LastAccessed should be `json:"last_accessed"`.

nshumoogum replied:
To add tests to check response data, you will want to write a reusable function in handler_test.go:

```
func unmarshalJSON(body io.Reader) (*session.Session, error) {
	b, err := ioutil.ReadAll(body)
	if err != nil {
		return nil, err
	}

	var session *session.Session
	if err = json.Unmarshal(b, &session); err != nil {
		return nil, err
	}

	return session, err
}
```

Call function and handle error like so:

```
...
    sessionResp, err := unmarshalJSON(resp.Body)
    if err != nil {
        t.Errorf("fail to retrieve response body")
    }
...
```

Then you can add the following assertions:

```
So(sessionResp.Email, ShouldEqual, "test@test.com")
So(sessionResp.ID, ShouldNotBeEmpty)
So(sessionResp.LastAccessed, ShouldHappenBetween, timeBefore, timeAfter)
So(sessionResp.Start, ShouldHappenBetween, timeBefore, timeAfter)
```

You can set the `timeBefore` as a global variable: `var timeBefore = time.Now()` for re-use across tests

And the `timeAfter` should be set in each test once the request has been made, for safety measure I would add 1 second to value, like so:

`timeAfter := time.Now().Add(1 * time.Second)`

At this point your tests will fail due to `LastAccessed` and `Start` variables in session object do not register the timezone correctly when reading it back out in the test (this maybe due to formatting the timestamp). I would use `time.Now().UTC()` in the Update function in session package to set a UTC time which will be understood in any system.

```
// Update updates the session with email parameter
func (s *Session) Update(email string) (*Session, error) {
	id, err := newID()
	if err != nil {
		return nil, err
	}

	return &Session{
		ID:           id,
		Email:        email,
		Start:        time.Now().UTC(),
		LastAccessed: time.Now().UTC(),
	}, nil
}
```

morses-code committed:
Fixed json tag for sessions, updated tests with new assertions

nshumoogum reviewed (approved):
👍🏼
